### DISCUSSION TITLE
Solving Inverse Kinematics on SE(3) (rotation part specifically)

### DISCUSSION THREAD

**Discussion by ThisIsBohao** [NVlabs/curobo#269] [Research Q&A] [answered]
Hi, I have a quick conceptual question on how curobo implements a differentiable representation of the end effector kinematics in order to solve inverse kinematics problems on SE(3) using optimization.
I noticed that curobo computes the 4x4 homogenous transformation matrix. It also provides functions that converts rotation matrices to quaternions. I wonder how curobo implements the gradient of the kinematics constraints so that the orientation eventually converges to target values. Is the gradient not computed explicitly but computed by automatic differentiation in torch library?

**Comment by balakumar-s** [comment, marked as answer]
It is calculated explicitly here: https://github.com/NVlabs/curobo/blob/911da8cb24949da599178061545a3338fbc082ea/src/curobo/curobolib/cpp/kinematics_fused_kernel.cu#L1037

We also refer to the calculation in Table 7 in our report (page 50): https://curobo.org/reports/curobo_report.pdf
<img width="556" alt="image" src="https://github.com/NVlabs/curobo/assets/3141191/5d489863-6ff0-43a0-9b08-b22ac2af387c">
